torch/distributed/_sharding_spec/api.py

```diff
@@ -0,0 +1,98 @@
+from abc import ABC
+import torch
+from typing import List, Union
+
+from ._internals import is_valid_device
+
+Device = Union[torch.device, int, str]
+
+class PlacementSpec(ABC):
+    """
+    Base class representing the placement of an entity.
+    """
+    pass
+
+class DevicePlacement(PlacementSpec):
+    """
+    Associates placement of an entity with a single device. The device can be a
+    local device or a remote device specified by one of the following remote
+    formats:
+
+        1. "rank:<rank>/<device>" (ex: "rank:0/cuda:0").
+        2. "<worker_name>/<device>" (ex: "trainer0/cuda:0").
+
+    Args:
+        device(str, :class:`torch.device`): The device to place the entity on.
+    """
+    def __init__(self, device: Device):
+        super(DevicePlacement, self).__init__()
+        if not is_valid_device(device):
+            raise ValueError(f'{device} is not a valid device')
+        self._device = device
+
+    @property
+    def device(self) -> Device:
+        """
+        Retrieves the device for placement.
+        """
+        return self._device
+
+class ChunkShardingSpec(PlacementSpec):
```

Comment:
ChunkShardingSpec , DevicePlacement and PlacementSpec - naming seems a bit off. Some have 'Placement', some have 'Spec' in their names. It's not obvious they belong to the same class hierarchy.